src/main/netplay.c

```diff
@@ -0,0 +1,618 @@
+        SDLNet_Write32((int32_t)size, &output_data[buffer_pos]); //file data size
+        buffer_pos += 4;
+        memcpy(&output_data[buffer_pos], data, size); //file data
+        buffer_pos += size;
+
+        SDLNet_TCP_Send(l_tcpSocket, &output_data[0], buffer_pos);
+    }
+    else
+    {
+        request = TCP_RECEIVE_SAVE;
+        memcpy(&output_data[buffer_pos], &request, 1);
+        ++buffer_pos;
+
+        //name of the file we are requesting
+        memcpy(&output_data[buffer_pos], short_filename, strlen(short_filename) + 1);
+        buffer_pos += strlen(short_filename) + 1;
+
+        SDLNet_TCP_Send(l_tcpSocket, &output_data[0], buffer_pos);
+        size_t recv = 0;
+        char *data_array = data;
+        while (recv < size)
+            recv += SDLNet_TCP_Recv(l_tcpSocket, data_array + recv, size - recv);
+
+        int sum = 0;
+        for (int i = 0; i < size; ++i)
+            sum |= data_array[i];
+
+        if (sum == 0) //all zeros means there is no save file
+            ret = file_open_error;
+        else
+            ret = file_ok;
+    }
+    free(output_data);
+    return ret;
+}
+
+void netplay_sync_settings(uint32_t *count_per_op, uint32_t *disable_extra_mem, int32_t *si_dma_duration, uint32_t *emumode, int32_t *no_compiled_jump)
+{
+    if (!netplay_is_init())
+        return;
+
+    char output_data[21];
+    uint8_t request;
+    if (l_netplay_control[0] != -1) //player 1 is the source of truth for settings
+    {
+        request = TCP_SEND_SETTINGS;
+        memcpy(&output_data[0], &request, 1);
+        SDLNet_Write32(*count_per_op, &output_data[1]);
+        SDLNet_Write32(*disable_extra_mem, &output_data[5]);
+        SDLNet_Write32(*si_dma_duration, &output_data[9]);
+        SDLNet_Write32(*emumode, &output_data[13]);
+        SDLNet_Write32(*no_compiled_jump, &output_data[17]);
+        SDLNet_TCP_Send(l_tcpSocket, &output_data[0], 21);
+    }
+    else
+    {
+        request = TCP_RECEIVE_SETTINGS;
+        memcpy(&output_data[0], &request, 1);
+        SDLNet_TCP_Send(l_tcpSocket, &output_data[0], 1);
+        int32_t recv = 0;
+        while (recv < 20)
+            recv += SDLNet_TCP_Recv(l_tcpSocket, &output_data[recv], 20 - recv);
+        *count_per_op = SDLNet_Read32(&output_data[0]);
+        *disable_extra_mem = SDLNet_Read32(&output_data[4]);
+        *si_dma_duration = SDLNet_Read32(&output_data[8]);
+        *emumode = SDLNet_Read32(&output_data[12]);
+        *no_compiled_jump = SDLNet_Read32(&output_data[16]);
+    }
+}
+
+void netplay_check_sync(struct cp0* cp0)
+{
+    //This function is used to check if games have desynced
+    //Every 60 VIs, it sends the value of the CP0 registers to the server
+    //The server will compare the values, and update the status byte if it detects a desync
+    if (!netplay_is_init())
+        return;
+
+    const uint32_t* cp0_regs = r4300_cp0_regs(cp0);
+
+    if (l_vi_counter % 60 == 0)
```

Comment:
Is `60` an universal value? Does it work for PAL games?

Reply:
Yes it'll work with PAL, it could be any number really, it's just there to make it transmit a desync check roughly every second.